alacritty/src/input/keyboard.rs

```diff
@@ -0,0 +1,584 @@
+    let modifiers = mods.into();
+
+    let kitty_seq = mode.intersects(
+        TermMode::REPORT_ALL_KEYS_AS_ESC
+            | TermMode::DISAMBIGUATE_ESC_CODES
+            | TermMode::REPORT_EVENT_TYPES,
+    );
+
+    let kitty_encode_all = mode.contains(TermMode::REPORT_ALL_KEYS_AS_ESC);
+    // When the key event represents a `Press` it'll be encoded as `1`, so we can omit it.
+    let kitty_event_type = mode.contains(TermMode::REPORT_EVENT_TYPES)
+        && (key.repeat || key.state == ElementState::Released);
+
+    let context =
+        SequenceBuildingContext { mode, modifiers, kitty_seq, kitty_encode_all, kitty_event_type };
+
+    let sequence_base = context
+        .try_from_numpad(&key)
+        .or_else(|| context.try_from_named(&key))
+        .or_else(|| context.try_from_control_char_or_mod(&key))
+        .or_else(|| context.try_from_textual(&key));
+
+    let (payload, terminator) = match sequence_base {
+        Some(SequenceBase { payload, terminator }) => (payload, terminator),
+        _ => return Vec::new(),
+    };
+
+    let mut payload = format!("\x1b[{}", payload);
+
+    // Add modifiers information.
+    if kitty_event_type
+        || !modifiers.is_empty()
+        || (mode.contains(TermMode::REPORT_ASSOCIATED_TEXT) && key.text.is_some())
+    {
+        payload.push_str(&format!(";{}", modifiers.encode_esc_sequence()));
+    }
+
+    // Push event type.
+    if kitty_event_type {
+        payload.push(':');
+        let event_type = match key.state {
+            _ if key.repeat => '2',
+            ElementState::Pressed => '1',
+            ElementState::Released => '3',
+        };
+        payload.push(event_type);
+    }
+
+    // Associated text is not reported when the control/alt/logo is pressesed.
+    if mode.contains(TermMode::REPORT_ASSOCIATED_TEXT)
+        && key.state != ElementState::Released
+        && (modifiers.is_empty() || modifiers == SequenceModifiers::SHIFT)
+    {
+        if let Some(text) = key.text {
+            let mut codepoints = text.chars().map(u32::from);
+            if let Some(codepoint) = codepoints.next() {
+                payload.push_str(&format!(";{codepoint}"));
+            }
+            for codepoint in codepoints {
+                payload.push_str(&format!(":{codepoint}"));
+            }
+        }
+    }
+
+    payload.push(terminator.encode_esc_sequence());
+
+    payload.into_bytes()
+}
+
+/// Helper to build escape sequence payloads from [`KeyEvent`].
+pub struct SequenceBuildingContext {
+    mode: TermMode,
+    /// The emitted sequence should follow the kitty keyboard protocol.
+    kitty_seq: bool,
+    /// Encode all the keys according to the protocol.
+    kitty_encode_all: bool,
+    /// Report event types.
+    kitty_event_type: bool,
+    modifiers: SequenceModifiers,
+}
+
+impl SequenceBuildingContext {
+    /// Try building sequence from the event's emitting text.
+    fn try_from_textual(&self, key: &KeyEvent) -> Option<SequenceBase> {
```

Comment:
I've just noticed this, but why do all these `try_from` impls not return `Self`?

Reply:
That's because we use `Context` to build common concepts, it's not really a builder. Alternatively we'd need to carry the value inside this builder, which is not good.

The design is simply to have common args passed across the functions. I had a version with `Self` but I wasn't sure about it, since it was less clear.

Reply:
It's just the name that doesn't make sense here, I'm not opposed to the mechanism.

A `try_from` method that takes `self` and doesn't return `Self` is just nonsense. Maybe just rename the struct to `SequenceBuilder` and make it `try_build_textual` etc?

Reply:
>A try_from method that takes self and doesn't return Self is just nonsense. Maybe just rename the struct to SequenceBuilder and make it try_build_textual etc?

I guess, I applied the suggestion.